### PR TITLE
Сonvert components to state components.

### PR DESCRIPTION
**Pull request by olha-khomiakova**
Сonvert components into state components to see with which markup some component will be displayed.
All tests passed.
ESLint passed.

### CODE REVIEW

**Review by dkalinin — APPROVED** [review]

File: js/app/components/component/index.js [hunk 1]
```diff
@@ -10,11 +10,18 @@ export class Component {
    */
   constructor(container) {
     this._container = container;
+    this.init();
     this.render();
     this.initNestedComponents();
     this.addEventListeners();
```
Comment (dkalinin) [inline js/app/components/component/index.js:16]: Please move this to init

**Review by dkalinin — CHANGES_REQUESTED** [review]

File: js/app/file-list-page/file-list-page/index.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ export class FileListPage extends Component {
     this._container = container;
     this._titleService = titleService;
     this._setTitle();
-    this._init();
+    this._initNestedComponentsWithDataFromServer().then();
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:24]: Let's use init method instead of overriding constructor.

**Review by dkalinin — CHANGES_REQUESTED** [review]

File: js/app/components/component/index.js (hunk 1 above)
Comment (olha-khomiakova) [reply js/app/components/component/index.js:16]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 2 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:24]: Done.

File: js/app/components/component/index.js [hunk 3]
```diff
@@ -7,19 +7,21 @@ export class Component {
...
+   * @param {Object} params
    */
-  constructor(container) {
+  constructor(container, ...params) {
     this._container = container;
-    this.init();
-    this.render();
-    this.initNestedComponents();
-    this.addEventListeners();
+    this.init(params);
```
Comment (dkalinin) [inline js/app/components/component/index.js:14]: You can use `this.init(...params)`

File: js/app/components/component/index.js [hunk 4]
```diff
@@ -7,19 +7,21 @@ export class Component {
...
 
   /**
    * Inits the elements that must be created before rendering the component.
+   * @param {Array} params
    */
-  init() {
+  init(params) {
+    this.render();
+    this.initNestedComponents();
+    this.addEventListeners();
```
Comment (dkalinin) [inline js/app/components/component/index.js]: Please read the documentation and adjust init behavior properly.

File: js/app/components/form/index.js [hunk 5]
```diff
@@ -46,7 +46,7 @@ export class Form extends Component {
    */
   set buttonText(value) {
     this._buttonText = value;
-    this.render();
+    this.init();
```
Comment (dkalinin) [inline js/app/components/form/index.js:49]: Init should be called once — during initialization of a component.

File: js/app/components/form/index.js [hunk 6]
```diff
@@ -18,23 +27,35 @@ export class Form extends Component {
    * @param {Function} inputsCreator is a function that inits inputs
    */
   initInputs(inputsCreator) {
+    this._inputsCreator = inputsCreator;
     inputsCreator(this.getElement('inputs'));
```
Comment (dkalinin) [inline js/app/components/form/index.js:31]: This should not be called here but should re-render the component.

**Review by dkalinin — CHANGES_REQUESTED** [review]
PTAL

File: js/app/components/form/index.js (hunk 6 above)
Comment (olha-khomiakova) [reply js/app/components/form/index.js:31]: Done.

File: js/app/components/form/index.js (hunk 5 above)
Comment (olha-khomiakova) [reply js/app/components/form/index.js:49]: Done.

File: js/app/components/component/index.js (hunk 4 above)
Comment (olha-khomiakova) [reply js/app/components/component/index.js]: Refactor.

File: js/app/components/component/index.js (hunk 3 above)
Comment (olha-khomiakova) [reply js/app/components/component/index.js:14]: Done.

File: js/app/services/dev-fetch/index.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import fetchMock from '../../../node_modules/fetch-mock/esm/client.js';
 
+/* eslint-disable no-debugger, no-console */
```
Comment (dkalinin) [inline js/app/services/dev-fetch/index.js:3]: Please remove no-debugger.

File: js/tests/app/file-list-page/file-list-page/index.js [hunk 8]
```diff
@@ -98,29 +99,28 @@ module('file-list-page', function() {
...
         return Promise.reject(new ServerErrorResponse('Server Error'));
       },
-      getFolder: (id) =>{
+      getFolder: (id) => {
         assert.strictEqual(id, '00000', 'ID should be the same as provided from the outside.');
         return Promise.reject(new ServerErrorResponse('Server Error'));
       },
-      userPreferences: () =>{
+      userPreferences: () => {
         return Promise.resolve({username: 'test', rootFolderId: '00000'});
```
Comment (dkalinin) [inline js/tests/app/file-list-page/file-list-page/index.js:111]: Let's replace '00000' with local variable.

**Review by dkalinin — CHANGES_REQUESTED** [review]
PTAL

File: js/app/components/component/index.js [hunk 9]
```diff
@@ -7,9 +7,18 @@ export class Component {
...
+   * @param {...Object} params
    */
-  constructor(container) {
+  constructor(container, ...params) {
     this._container = container;
+    this.init(...params);
+  }
+
+  /**
+   * Initializes the component and provides the required services.
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/components/component/index.js:18]: Done.

File: js/app/file-list-page/breadcrumbs/index.js [hunk 10]
```diff
@@ -34,7 +42,8 @@ export class Breadcrumbs extends Component {
                         <strong class="path cropped">
                             <a href="#" title="Return to previous directory"> <span class="glyphicon glyphicon-level-up"
                                                                                     aria-hidden="true"></span></a>
-                            <span data-fh="crumbs"></span>
+                            <span data-fh="crumbs" ${this._iconClass || `class = "${this._classForErrorMessage}"`}>
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/breadcrumbs/index.js:45]: Changed this._iconClass to this._classForErrorMessage.

File: js/app/file-list-page/file-list-page/index.js [hunk 11]
```diff
@@ -47,55 +57,39 @@ export class FileListPage extends Component {
...
-
-    this.apiService.userPreferences().then((userPreferences) => {
-      this._userDetails.username = userPreferences.name;
-
-      this.apiService.getFolder(userPreferences.rootFolderId).then((folder) => {
-        this._breadcrumbs.addBreadcrumb(folder.name);
-      }).catch((error) => {
-        this._breadcrumbs.errorMessage = 'Unable to load';
-      });
+  _initNestedComponentsWithDataFromServer() {
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:74]: Done.

File: js/app/services/dev-fetch/index.js (hunk 7 above)
Comment (olha-khomiakova) [reply js/app/services/dev-fetch/index.js:3]: Done.

File: js/app/file-list-page/file-list-item/index.js [hunk 12]
```diff
@@ -11,52 +11,48 @@ export class FileListItem extends Component {
...
-    this.name = fileListModel.name;
-    this.mimeType = fileListModel.mimeType;
-    this.size = fileListModel.size;
-    this._initMarkup();
+  }
+
+  /**
+   * @inheritDoc
+   */
+  init(...params) {
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-item/index.js:21]: Done.

File: js/app/file-list-page/file-list-item/index.js [hunk 13]
```diff
@@ -11,52 +11,48 @@ export class FileListItem extends Component {
...
   constructor(container, fileListModel) {
-    super(container);
+    super(container, fileListModel);
     this._container = container;
-    this.id = fileListModel.id;
-    this.name = fileListModel.name;
-    this.mimeType = fileListModel.mimeType;
-    this.size = fileListModel.size;
-    this._initMarkup();
+  }
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-item/index.js:16]: Done.

File: js/app/file-list-page/file-list-item/index.js [hunk 14]
```diff
@@ -68,11 +64,12 @@ export class FileListItem extends Component {
...
          <td class="first-column"></td>
-         <td data-fh="glyphicon" class="glyphicon second-column" aria-hidden="true"></td>
+         <td data-fh="glyphicon" class="glyphicon second-column ${glyphicon[this._mimeType]}" 
+         aria-hidden="true"></td>
          <td class="third-column">
-             <a data-fh="link" class="content-name cropped" href="#"></a>
+             <a data-fh="link" class="content-name cropped" href="#">${this._name}</a>
          </td>
-         <td data-fh="size" class="fourth-column"></td>
+         <td data-fh="size" class="fourth-column">${this._convertSize()}</td>
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-item/index.js:72]: Done.

File: js/tests/app/file-list-page/file-list-page/index.js (hunk 8 above)
Comment (olha-khomiakova) [reply js/tests/app/file-list-page/file-list-page/index.js]: Done.

File: js/app/file-list-page/file-list/index.js [hunk 15]
```diff
@@ -12,25 +12,23 @@ export class FileList extends Component {
    * @param {ListItemModel} list
    */
   addList(list) {
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:14]: Let's replace with setter for "items" field.

File: js/app/file-list-page/file-list/index.js [hunk 16]
```diff
@@ -12,25 +12,23 @@ export class FileList extends Component {
...
+    this._list = list;
+    this._message = null;
+    this._classForErrorMessage = null;
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:17]: Why are we doing it here? They have their setters.

File: js/app/file-list-page/file-list/index.js [hunk 17]
```diff
@@ -12,25 +12,23 @@ export class FileList extends Component {
...
-      }
-    });
   }
+
   /**
    * Sets an error massage.
    * @param {String} errorMessage
    */
   set errorMessage(errorMessage) {
     this._message = 'Unable to load a file list.';
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:26]: How about `errorMessage`?

File: js/app/file-list-page/file-list/index.js [hunk 18]
```diff
@@ -12,25 +12,23 @@ export class FileList extends Component {
...
     this._classForErrorMessage = 'error-message';
+    this._list = null;
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:28]: Please remove. It's not a responsibility of errorMessage setter.

File: js/app/file-list-page/file-list/index.js [hunk 19]
```diff
@@ -50,6 +48,22 @@ export class FileList extends Component {
     return -1;
   }
 
+  /**
+   * @inheritDoc
+   */
+  initNestedComponents() {
+    if (this._list) {
+      this._list.listItemModel.sort(this._sortByTypeAndName);
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:56]: Why not `this._list.sort(...)` ?

File: js/app/file-list-page/link-with-action/index.js [hunk 20]
```diff
@@ -9,7 +9,7 @@ export class LinkWithActions extends Component {
    */
   init(...params) {
     this._text = '';
-    super.init(...params);
+    super.init();
```
Comment (dkalinin) [inline js/app/file-list-page/link-with-action/index.js]: ...params are ignored.
Are they needed?

File: js/app/login/login-form/index.js [hunk 21]
```diff
@@ -13,7 +13,7 @@ export class LoginForm extends Component {
   init(...param) {
     this._loginValue = '';
     this._passwordValue = '';
-    super.init(...param);
+    super.init();
```
Comment (dkalinin) [inline js/app/login/login-form/index.js:16]: ...params ignored.

File: js/app/registration/registration-form/index.js [hunk 22]
```diff
@@ -14,7 +14,7 @@ export class RegistrationForm extends Component {
     this._loginValue = '';
     this._passwordValue = '';
     this._confirmPasswordValue = '';
-    super.init(...params);
+    super.init();
```
Comment (dkalinin) [inline js/app/registration/registration-form/index.js:17]: ...params ignored.

File: js/tests/app/file-list-page/breadcrumbs/index.js [hunk 23]
```diff
@@ -2,13 +2,16 @@ import {Breadcrumbs} from '../../../../app/file-list-page/breadcrumbs';
 
 const {module, test} = QUnit;
 module('breadcrumbs', function() {
-  test('Should be breadcrumbs rendered with certain data.', function(assert) {
+  test('Should render breadcrumbs with certain data.', function(assert) {
```
Comment (dkalinin) [inline js/tests/app/file-list-page/breadcrumbs/index.js:5]: iconClass field is not tested. Please check all public fields of each component.

**Review by dkalinin — CHANGES_REQUESTED** [review]

File: js/app/components/form/index.js [hunk 24]
```diff
@@ -6,12 +6,11 @@ import {Component} from '../component';
  */
 export class Form extends Component {
   /**
-   * Returns a new instance of the Form.
-   * @param {Element} container in which form will be rendered
+   * @inheritDoc
    */
-  constructor(container) {
-    super(container);
+  init(...params) {
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/components/form/index.js:11]: Done.

File: js/app/components/validatable-input/index.js [hunk 25]
```diff
@@ -5,21 +5,13 @@ import {createAUniqueId} from '../../services/unique-id';
...
-    this._id = createAUniqueId();
-  }
-
   /**
    * @inheritDoc
    */
   init(...params) {
     this._message = '';
     this._inputValue = '';
+    this._id = createAUniqueId();
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/components/validatable-input/index.js:14]: Done.

File: js/app/file-list-page/breadcrumbs/index.js [hunk 26]
```diff
@@ -10,7 +10,7 @@ export class Breadcrumbs extends Component {
   init(...params) {
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/breadcrumbs/index.js:10]: Done.

File: js/app/file-list-page/file-list/index.js (hunk 15 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js:14]: Done.

File: js/app/file-list-page/file-list/index.js (hunk 16 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js:17]: Done.

File: js/app/file-list-page/file-list/index.js (hunk 18 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js:28]: Done.

File: js/app/file-list-page/file-list/index.js (hunk 19 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js:56]: Done.

File: js/app/file-list-page/link-with-action/index.js (hunk 20 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/link-with-action/index.js]: Done.

File: js/app/login/login-form/index.js (hunk 21 above)
Comment (olha-khomiakova) [reply js/app/login/login-form/index.js:16]: Done.

File: js/app/registration/registration-form/index.js (hunk 22 above)
Comment (olha-khomiakova) [reply js/app/registration/registration-form/index.js:17]: Done.

File: js/app/file-list-page/breadcrumbs/index.js [hunk 27]
```diff
@@ -42,8 +42,8 @@ export class Breadcrumbs extends Component {
                         <strong class="path cropped">
                             <a href="#" title="Return to previous directory"> <span class="glyphicon glyphicon-level-up"
                                                                                     aria-hidden="true"></span></a>
-                            <span data-fh="crumbs" ${this._iconClass || `class = "${this._classForErrorMessage}"`}>
-                            ${this._breadcrumbs}</span>
+                            <span data-fh="crumbs" ${this._classForErrorMessage &&
+                            `class = "${this._classForErrorMessage}"`}>${this._breadcrumbs}</span>
```
[the remark this answers is not in the document]
Comment (olha-khomiakova) [reply js/app/file-list-page/breadcrumbs/index.js:46]: Done.

File: js/app/file-list-page/file-list/index.js (hunk 17 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js:26]: Done.

File: js/tests/app/file-list-page/breadcrumbs/index.js (hunk 23 above)
Comment (olha-khomiakova) [reply js/tests/app/file-list-page/breadcrumbs/index.js:5]: I haven`t iconClass field in Breadcrumbs.

File: js/app/file-list-page/file-list/index.js [hunk 28]
```diff
@@ -7,14 +7,19 @@ import {FileListModel, FolderListModel} from '../list-item-model';
  * Component that renders a FileList.
  */
 export class FileList extends Component {
+  /**
+   * @inheritDoc
+   */
+  init() {
+    // this._message = '';
+  }
```
Comment (dkalinin) [inline js/app/file-list-page/file-list/index.js:15]: Please remove.

**Review by dkalinin — APPROVED** [review]
Please fix and merge.

File: js/app/file-list-page/file-list/index.js (hunk 28 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list/index.js]: Done.

**Pull request merged**